File: solana-monitoring/fetch/fetch_full_meta_data_transaction_details_per_transaction_for_particular_wallet.py
import json, argparse, requests, time

from pathlib import Path

parser = argparse.ArgumentParser()
parser.add_argument("input_folder", help="Folder containing signature JSON files")
parser.add_argument("--output_folder", default="full_transaction_details_per_wallet")
args = parser.parse_args()

input_path = Path(args.input_folder)
output_path = Path(args.output_folder)
output_path.mkdir(exist_ok=True)

files = sorted(input_path.glob("*.json"))

RPC_URL = "https://api.devnet.solana.com"
HEADERS = {"Content-Type": "application/json"}
results = []

for file in files:
    print(f"\n📂 Processing {file.name}")
    with open(file) as f: txs = json.load(f)
    signatures = [tx["signature"] for tx in txs if "signature" in tx]
    results = []

    for i, sig in enumerate(signatures):
        print(f"[{i+1}/{len(signatures)}] {sig}")
        try:
            res = requests.post(RPC_URL, headers=HEADERS, json={
                "jsonrpc": "2.0", "id": 1, "method": "getTransaction",
                "params": [sig, {"encoding": "jsonParsed"}]
            }, timeout=10).json().get("result")
            if res:
                results.append({
                    "signature": sig,
                    "blockTime": res.get("blockTime"),
                    "slot": res.get("slot"),
                    "fee": res["meta"]["fee"],
                    "status": res["meta"].get("err"),
                    "preBalances": res["meta"]["preBalances"],
                    "postBalances": res["meta"]["postBalances"],
                    "tokenTransfers": res["meta"].get("postTokenBalances", []),
                    "logMessages": res["meta"].get("logMessages"),
                    "instructions": res["transaction"]["message"].get("instructions", []),
                    "accountKeys": [x["pubkey"] if isinstance(x, dict) else x for x in res["transaction"]["message"]["accountKeys"]]
                })
            time.sleep(0.25)
        except Exception as e:
            print(f"❌ {e}")

    out_file = output_path / f"transactionsmetadata_{file.stem}.json"
    with open(out_file, "w") as f: json.dump(results, f, indent=2)
    print(f"✅ Saved {len(results)} txs to {out_file}")

File: solana-monitoring/fetch/fetch_full_meta_data_transaction_details_per_transaction_for_particular_wallet.js
#!/usr/bin/env node

const fs = require('fs');
const fetch = require('node-fetch');
const path = require('path');
const minimist = require('minimist');

const args = minimist(process.argv.slice(2));
const inputFolder = args._[0];
const outputFolder = args.output_folder || 'full_transaction_details_per_wallet';

if (!inputFolder) {
  console.log('Usage: node fetch_full_meta_data_transaction_details_per_transaction_for_particular_wallet.js <input_folder> [--output_folder <output_folder>]');
  process.exit(1);
}

const inputPath = path.resolve(inputFolder);
const outputPath = path.resolve(outputFolder);
if (!fs.existsSync(outputPath)) fs.mkdirSync(outputPath, { recursive: true });

const files = fs.readdirSync(inputPath).filter(f => f.endsWith('.json')).sort();
const RPC_URL = 'https://api.devnet.solana.com';
const HEADERS = { 'Content-Type': 'application/json' };

(async () => {
  for (const file of files) {
    console.log(`\n📂 Processing ${file}`);
    const txs = JSON.parse(fs.readFileSync(path.join(inputPath, file), 'utf8'));
    const signatures = txs.filter(tx => tx.signature).map(tx => tx.signature);
    const results = [];
    for (let i = 0; i < signatures.length; i++) {
      const sig = signatures[i];
      console.log(`[${i + 1}/${signatures.length}] ${sig}`);
      try {
        const payload = {
          jsonrpc: '2.0', id: 1, method: 'getTransaction',
          params: [sig, { encoding: 'jsonParsed' }]
        };
        const res = await fetch(RPC_URL, {
          method: 'POST',
          headers: HEADERS,
          body: JSON.stringify(payload),
          timeout: 10000
        });
        const data = await res.json();
        const result = data.result;
        if (result) {
          results.push({
            signature: sig,
            blockTime: result.blockTime,
            slot: result.slot,
            fee: result.meta.fee,
            status: result.meta.err,
            preBalances: result.meta.preBalances,
            postBalances: result.meta.postBalances,
            tokenTransfers: result.meta.postTokenBalances || [],
            logMessages: result.meta.logMessages,
            instructions: result.transaction.message.instructions || [],
            accountKeys: (result.transaction.message.accountKeys || []).map(x => (typeof x === 'object' ? x.pubkey : x))
          });
        }
        await new Promise(r => setTimeout(r, 250));
      } catch (e) {
        console.log(`❌ ${e}`);
      }
    }
    const outFile = path.join(outputPath, `transactionsmetadata_${path.parse(file).name}.json`);
    fs.writeFileSync(outFile, JSON.stringify(results, null, 2));
    console.log(`✅ Saved ${results.length} txs to ${outFile}`);
  }
})(); 